### DISCUSSION TITLE
Should rate-limited tasks increment the prefetch limit?

### DISCUSSION THREAD

**Discussion by RJPercival** [celery/celery#7803] [Q&A]
When Celery is configured with a prefetch limit of 1, and receives a task with an ETA but that ETA is in the future, Celery will increment the prefetch limit and schedule that task to be run when its ETA elapses. The incremented prefetch limit allows the worker to fetch another task while holding the future-dated task. This seems sensible.

The behaviour is different for rate-limited tasks. If a worker receives a task but there are not sufficient tokens (due to rate-limiting) to execute that task, it will similarly schedule that task to be run when it is expected that a token will be available, but the prefetch limit is **not** incremented. This means the worker is unable to fetch any more tasks and sits idle. It does so even though the next task it could fetch might not be rate-limited. This seems like a bug - should the prefetch limit not also be incremented in this scenario?

Here is the relevant code:
https://github.com/celery/celery/blob/febe6956fabecdbce0150757a536b2006bd36407/celery/worker/strategy.py#L190-L202

**Comment by auvipy** [comment]
can you share a proof of concept of what you are thinking? with a failing test?

**Reply by RJPercival** [reply]
Essentially, I'm wondering whether `consumer.qos.increment_eventually()` should be called on line 202. Unfortunately there doesn't seem to be much in the way of existing tests for rate-limiting, nor for workers configured with a prefetch limit of 1. I don't currently have the time to add those tests without some guidance as to whether this behaviour change would be a welcome one.

**Reply by RJPercival** [reply]
As a sponsor (I'm a software engineer at @kraken-tech), would it be possible to get some priority support on this please?

**Reply by auvipy** [reply]
@thedrow are you/ your team up for a gig?

**Reply by RJPercival** [reply]
@auvipy / @thedrow: Any chance of getting some support on this in the near future? I'm happy for that to take the form of some pairing where someone talks me through what the fix and tests should look like.

**Reply by RJPercival** [reply]
@auvipy, I guess not?

**Reply by auvipy** [reply]
can you shoot me a line on auvipy@gmail.com please?

**Reply by RJPercival** [reply]
Sorry @auvipy, I missed this response. I'll email you now.

**Comment by auvipy** [comment]
@RJPercival can you please try incrementing this and see what happens?

**Reply by RJPercival** [reply]
There don't appear to be any tests for this, so there's no straightforward way to "see what happens". That was one of things I was looking for support with - getting some tests written that cover rate-limiting comprehensively.

**Comment by RJPercival** [comment]
Related issue: https://github.com/celery/celery/issues/8119

**Comment by dbartenstein** [comment]
Has there been any progress on this?

**Reply by RJPercival** [reply]
Not that I'm aware of. I haven't been able to get any support from the Celery maintainers unfortunately.
